packages/drawer/src/mwc-drawer.ts

```diff
@@ -0,0 +1,166 @@
+  open(): void;
+  close(): void;
+}
+
+export declare var DrawerFoundation: {
+  prototype: DrawerFoundation;
+  new(adapter: Adapter): DrawerFoundation;
+}
+
+@customElement('mwc-drawer' as any)
+export class Drawer extends BaseElement {
+
+  @query('.mdc-drawer')
+  mdcRoot!: HTMLElement;
+
+  protected mdcFoundation: MDCDismissibleDrawerFoundation|MDCModalDrawerFoundation|undefined;
+
+  protected get mdcFoundationClass(): (typeof DrawerFoundation) {
+    if (this.type === 'modal') {
+      return MDCModalDrawerFoundation;
+    } else {
+      return MDCDismissibleDrawerFoundation;
+    }
+  }
+
+  protected createAdapter() {
+    return {
+      ...super.createAdapter(),
+      elementHasClass: (element, className) => element.classList.contains(className),
+      computeBoundingRect: () => this.mdcRoot.getBoundingClientRect(),
+      saveFocus: () => {
+        this._previousFocus = document.activeElement as HTMLElement;
+      },
+      restoreFocus: () => {
+        const previousFocus = this._previousFocus && this._previousFocus.focus;
+        if (this.mdcRoot.contains(document.activeElement) && previousFocus) {
+          this._previousFocus!.focus();
+        }
+      },
+      // TODO(sorvell): List integration like this may not work. Need to understand
+      // why this is here.
+      focusActiveNavigationItem: () => {
+        // const activeNavItemEl = this._root.querySelector(`.${MDCListFoundation.cssClasses.LIST_ITEM_ACTIVATED_CLASS}`)!;
+        // if (activeNavItemEl) {
+        //   (activeNavItemEl as HTMLElement).focus();
+        // }
+      },
+      notifyClose: () => {
+        this.open = false;
+        this.dispatchEvent(new Event(strings.CLOSE_EVENT, {bubbles: true, cancelable: true}))
+      },
+      notifyOpen: () => {
+        this.open = true;
+        this.dispatchEvent(new Event(strings.OPEN_EVENT, {bubbles: true, cancelable: true}))
+      },
+      trapFocus: () => {/*this._focusTrap.activate()*/},
+      releaseFocus: () => {/*this._focusTrap.deactivate()*/},
+    }
+  }
+
+  // TODO(sorvell): integrate focus trapping.
+  //private _focusTrap = undefined;
+  private _previousFocus: HTMLElement|undefined = undefined;
+
+  @observer(function(this: Drawer, value: boolean) {
+    if (this.type === '') {
+      return;
+    }
+    if (value) {
+      this.mdcFoundation.open();
+    } else {
+      this.mdcFoundation.close();
+    }
+  })
+  @property({type: Boolean, reflect: true})
+  open = false;
+
+  @property({type: Boolean})
+  hasHeader = false;
+
+  @property({reflect: true})
+  type = '';
+
+  renderStyle() {
+    return style;
+  }
+
+  render() {
+    const dismissible = this.type === 'dismissible' || this.type === 'modal';
+    const modal = this.type === 'modal';
+    return html`
+      ${this.renderStyle()}
+      <aside class="mdc-drawer
+          ${classMap({'mdc-drawer--dismissible': dismissible, 'mdc-drawer--modal': modal})}">
+        ${this.hasHeader ? html`
```

Comment:
Can we have this pulled out and make the ternary simpler?

Reply:
Done